projects/packages/publicize/src/jetpack-social-settings/class-settings.php

```diff
@@ -0,0 +1,251 @@
+	 * together with the auto conversion settings.
+	 *
+	 * TODO: Work out if this is possible on plugin upgrade
+	 *
+	 * @return void
+	 */
+	private function migrate_old_option() {
+		$auto_conversion_settings = get_option( 'jetpack_social_settings' );
+		if ( ! empty( $auto_conversion_settings['image'] ) ) {
+			update_option( self::OPTION_PREFIX . self::AUTOCONVERT_IMAGES, array( 'enabled' => $auto_conversion_settings['image'] ) );
+			delete_option( 'jetpack_social_settings' );
+		}
+
+		$sig_settings = get_option( 'jetpack_social_image_generator_settings' );
+		if ( isset( $sig_settings['defaults']['template'] ) ) {
+			update_option(
+				self::OPTION_PREFIX . self::IMAGE_GENERATOR_SETTINGS,
+				array(
+					'enabled'  => $sig_settings['enabled'],
+					'template' => $sig_settings['defaults']['template'],
+				)
+			);
+		}
+	}
+
+	/**
+	 * Register the settings.
+	 *
+	 * @return void
+	 */
+	public function register_settings() {
+		register_setting(
+			'general',
+			self::OPTION_PREFIX . self::AUTOCONVERT_IMAGES,
+			array(
+				'default'      => array(
+					'enabled' => true,
+				),
+				'show_in_rest' => array(
+					'schema' => array(
+						'type'       => 'object',
+						'properties' => array(
+							'enabled' => array(
+								'type' => 'boolean',
+							),
+						),
+					),
+				),
+				'type'         => 'boolean',
+			)
+		);
+
+		register_setting(
+			'general',
+			self::OPTION_PREFIX . self::IMAGE_GENERATOR_SETTINGS,
+			array(
+				'type'         => 'object',
+				'default'      => array(
+					'enabled'  => false,
+					'template' => Templates::DEFAULT_TEMPLATE,
+				),
+				'show_in_rest' => array(
+					'schema' => array(
+						'type'       => 'object',
+						'properties' => array(
+							'enabled'  => array(
+								'type' => 'boolean',
+							),
+							'template' => array(
+								'type' => 'string',
+							),
+						),
+					),
+				),
+			)
+		);
+
+		add_filter( 'rest_pre_update_setting', array( $this, 'update_settings' ), 10, 3 );
+	}
+
+	/**
+	 * Get the current settings.
+	 *
+	 * @param bool $with_available Whether to include the available status of the features.
+	 *
+	 * @return array
+	 */
+	public function get_settings( $with_available = false ) {
+		$this->migrate_old_option();
+
+		$settings = array(
+			'autoConversionSettings'       => get_option( self::OPTION_PREFIX . self::AUTOCONVERT_IMAGES, self::DEFAULT_AUTOCONVERT_IMAGES_SETTINGS ),
+			'socialImageGeneratorSettings' => get_option( self::OPTION_PREFIX . self::IMAGE_GENERATOR_SETTINGS, self::DEFAULT_IMAGE_GENERATOR_SETTINGS ),
+		);
+
+		// The feature cannot be enabled without Publicize.
+		if ( ! ( new Modules() )->is_active( 'publicize' ) ) {
+			$settings['autoConversionSettings']['enabled']       = false;
+			$settings['socialImageGeneratorSettings']['enabled'] = false;
+		}
+
+		if ( $with_available ) {
+			$settings['autoConversionSettings']['available']       = $this->is_auto_conversion_available();
```

Comment:
I was giving this a test, went to the Social admin page, and got this error:
```
Cannot access offset of type string on string in /usr/local/src/jetpack-monorepo/projects/packages/publicize/src/jetpack-social-settings/class-settings.php:144
```

My guess is that the options on my site are in a weird state, but we should maybe check for that. The return of `get_option` might not be an array.

Reply:
Added some safety checks, I tested it with clean state, then with the one someone would have coming from trunk